packages/sanity/src/structure/panes/documentList/ColumnsControl.tsx

```diff
@@ -0,0 +1,81 @@
+/* eslint-disable @sanity/i18n/no-attribute-string-literals */
+/* eslint-disable i18next/no-literal-string */
+import {Box, Button, Card, Checkbox, Flex, Menu, MenuButton, Stack, Text} from '@sanity/ui'
+import {type Column, type Table} from '@tanstack/react-table'
+import {useCallback} from 'react'
+import {type SanityDocument} from 'sanity'
+
+import {VISIBLE_COLUMN_LIMIT} from './useDocumentSheetColumns'
+
+type Props = {
+  table: Table<SanityDocument>
+}
+
+export function ColumnsControl({table}: Props) {
+  const isVisibleLimitReached =
+    table.getVisibleLeafColumns().filter((col) => col.getCanHide()).length >= VISIBLE_COLUMN_LIMIT
```

Comment:
`table` is from tanstack table and is already a reference to the instance, so memoising `isVisibleLimitReached` isn't possible, it should be rerun on every rerender, so that any changes to column visibility are detected